queue/src/main/resources/org/killbill/bus/dao/PersistentBusSqlDao.sql.stg

```diff
@@ -18,6 +18,13 @@ readyOrderByClause() ::= <<
     record_id asc
 >>
 
+reapWhereClause() ::= <<
```

Comment:
Isn't there a race condition where the same entries will be re-dispatched over and over again, until they are processed? In other words, if the polling of the bus is slower than the polling of the reaper, the same re-dispatched entries will be seen as old again, and re-dispatched.

The only option around it might be to update the `created_date` (cc @sbrossie).

Reply:
@pierre Agreed

Couple of options:
1. We configuring our reaping delay large enough but there is no guarantee : If there is a backup, potentially we end up in that state, and that will only make the problem worst
2. We change the design: Each reaper threads checks that 'things are not moving' for each other node before trying to re-balance entries. Since node don't know each other, that would require keeping a `last_record_id` processed for each existing nodes; i think this is feasible but adds some complexity ....

Reply:
How about a new `reaped_date` defaulted to null?.  Every time an entry is reaped the date is set. If the reaped date is not null then it could be reap.  That way we could conserve the created_date and know that it was reaped.

Reply:
Yeah, I was thinking hard to find a solution without changing the schema: `bus_events_history` is usually massive (hundreds of millions of rows) in large deployments, so changing it is not always an option. But I don't think there is a reliable way to handle this race condition otherwise.

On the plus side, if we end up changing it, we could add more information as suggested by @sbrossie above.

That being said, I think looking at `reaped_date != NULL` by itself isn't enough. Consider the scenario where a node reaps an entry, updates `reaped_date` to `now()` and then crashes before processing that reaped entry... 

@wwjbatista Let me have a chat with @sbrossie before making any change on this specifically (you can still go ahead and address the other comments) and we'll update here with how complicated we want this piece of code to be...

Reply:
@wwjbatista Here is what we are going to do... In a transaction:

* Move the current entry into the history table, but update the state to `REAPED` (`PersistentQueueEntryLifecycleState` to be updated)
* Insert a new entry into the main table, which is a copy of the previous entry, but with an updated `created_date`